### DISCUSSION TITLE
Tracking stage channels

### DISCUSSION THREAD

**Discussion by StateGovernment** [bwmarrin/discordgo#1176] [Q&A] [answered]
Hello community and devs, I am trying to develop a tracker that could track AMA events on a stage voice channel for my community on discord. Getting voice states for people who entered and exited the stage channel was possible through VoiceStateUpdate event but is there a way to get a list of current users in a voice channel?

**Comment by topi314** [comment, marked as answer]
to get a list of users in the voice channel make sure you have the voice state cache enabled see [here](https://github.com/bwmarrin/discordgo/blob/9e0783c37f5d254d9eb0139f5054db484f13e26a/state.go#L48) and then you can iterate over all voice states in the guild filtering for the `channelID`

**Reply by StateGovernment** [reply]
Hello @TopiSenpai Thank you for the suggestion,

I've tried retrieving the current State object from session and i can see that TrackVoice is enabled by default.
`fmt.Println(session.State.TrackVoice)`

But then when I tried to check for the current voice states from the guild, the result is empty.
`g, _ := session.Guild("")
fmt.Println(g.VoiceStates)`

Am I missing something here? Please help me out.

**Reply by topi314** [reply]
Make sure you have the guild voice states intent enabled

**Reply by StateGovernment** [reply]
@TopiSenpai 

`session.Identify.Intents = discordgo.IntentsGuildVoiceStates` I pass this as a command to enable voice states intent which works for VoiceStates Event but still Guild.VoiceStates is empty.

**Reply by topi314** [reply]
Are you able to provide some code of yours?

**Reply by StateGovernment** [reply]
@TopiSenpai Sure thing.

```go
func ama_tracker_init(session *discordgo.Session) {
	session.Identify.Intents = discordgo.IntentsGuildVoiceStates
	fmt.Println(session.Identify.Intents)

	g, _ := session.Guild(ZionGuildId)
	fmt.Println(g.VoiceStates)

	// Register the ReactionCreate func as a callback for ReactionCreate events.
	session.AddHandler(AmaStageCreate)
	// session.Identify.Intents = discordgo.IntentsGuildVoiceStates

	// Open a websocket connection to Discord and begin listening.
	err := session.Open()
	if err != nil {
		fmt.Println("error opening connection,", err)
		return
	}

	// Wait here until CTRL-C or other term signal is received.
	fmt.Println("Bot is now running. Press CTRL-C to exit.")
	sc := make(chan os.Signal, 1)
	signal.Notify(sc, syscall.SIGINT, syscall.SIGTERM, os.Interrupt)
	<-sc

	// Cleanly close down the Discord session.
	session.Close()
}    
 ```

Here you go. The event handling is working totally fine but couldn't retrieve a list of voice states live from the server.

**Reply by FedorLap2006** [reply]
`VoiceStates` are empty, because `VoiceStates`, as everything else in the cache, is only filled in after the session is `Open`ed, as cache relies on the gateway connection to get the data.

Additionally I would not recommend using `=` in the intents, as `Intents` is a bitmap and you can easily disable the necessary ones.
```diff
- session.Identify.Intents = discordgo.IntentsGuildVoiceStates
+ session.Identify.Intents |= discordgo.IntentsGuildVoiceStates
```

**Reply by StateGovernment** [reply]
@FedorLap2006 Thank you for the response, I've changed my code according, but still voicestates comes out empty.

```
func ama_tracker_init(session *discordgo.Session) {

	// Register the ReactionCreate func as a callback for ReactionCreate events.
	session.AddHandler(AmaStageCreate)
	// session.Identify.Intents = discordgo.IntentsGuildVoiceStates

	// Open a websocket connection to Discord and begin listening.
	err := session.Open()
	if err != nil {
		fmt.Println("error opening connection,", err)
		return
	}

	g, _ := session.Guild(ZionGuildId)
	fmt.Println(g.VoiceStates)

	session.Identify.Intents |= discordgo.IntentsGuildVoiceStates

	// Wait here until CTRL-C or other term signal is received.
	fmt.Println("Bot is now running. Press CTRL-C to exit.")
	sc := make(chan os.Signal, 1)
	signal.Notify(sc, syscall.SIGINT, syscall.SIGTERM, os.Interrupt)
	<-sc

	// Cleanly close down the Discord session.
	session.Close()
}
```

**Reply by topi314** [reply]
you need to set your intents before calling open

**Reply by FedorLap2006** [reply]
> @FedorLap2006 Thank you for the response, I've changed my code according, but still voicestates comes out empty.
> 
> ```
> func ama_tracker_init(session *discordgo.Session) {
> 
> 	// Register the ReactionCreate func as a callback for ReactionCreate events.
> 	session.AddHandler(AmaStageCreate)
> 	// session.Identify.Intents = discordgo.IntentsGuildVoiceStates
> 
> 	// Open a websocket connection to Discord and begin listening.
> 	err := session.Open()
> 	if err != nil {
> 		fmt.Println("error opening connection,", err)
> 		return
> 	}
> 
> 	g, _ := session.Guild(ZionGuildId)
> 	fmt.Println(g.VoiceStates)
> 
> 	session.Identify.Intents |= discordgo.IntentsGuildVoiceStates
> 
> 	// Wait here until CTRL-C or other term signal is received.
> 	fmt.Println("Bot is now running. Press CTRL-C to exit.")
> 	sc := make(chan os.Signal, 1)
> 	signal.Notify(sc, syscall.SIGINT, syscall.SIGTERM, os.Interrupt)
> 	<-sc
> 
> 	// Cleanly close down the Discord session.
> 	session.Close()
> }
> ```

First issue with this code is that you're specifiying intents after opening the session (as @TopiSenpai mentioned), but session needs to know the intents before creating a gateway connection, to send the correct `identify` packet.
Second issue is, that you're using `VoiceStates` still too soon, the voice state for a particular guild will be only filled after the `GuildCreate` event, as it returns the guild with all it's fields. Also, you can't use this in the `Ready` event handler, as the `Ready` event only contains unavailable guild objects (`{"id": "...", "unavailable": true}`), which will be lazily loaded by `GuildCreate`.

**Reply by StateGovernment** [reply]
@TopiSenpai @FedorLap2006 From what I understand, I've gotta declare intents before opening connection then read Guild Voicestates through Ready event which lazily fills. Changed the code accordingly still hasn't gotten any results on Voicestates.

```
func ama_tracker_init(session *discordgo.Session) {

	session.Identify.Intents |= discordgo.IntentsGuildVoiceStates

	// Register the ReactionCreate func as a callback for VoiceState events.
	// session.AddHandler(AmaStageCreate)
	session.AddHandler(GuildReady)
	// session.Identify.Intents = discordgo.IntentsGuildVoiceStates

	// Open a websocket connection to Discord and begin listening.
	err := session.Open()
	if err != nil {
		fmt.Println("error opening connection,", err)
		return
	}

	g, _ := session.Guild(ZionGuildId)
	fmt.Println(g.VoiceStates)

	// Wait here until CTRL-C or other term signal is received.
	fmt.Println("Bot is now running. Press CTRL-C to exit.")
	sc := make(chan os.Signal, 1)
	signal.Notify(sc, syscall.SIGINT, syscall.SIGTERM, os.Interrupt)
	<-sc

	// Cleanly close down the Discord session.
	session.Close()
}

func GuildReady(s *discordgo.Session, r *discordgo.Ready) {
	for _, i := range r.Guilds {
		defer fmt.Println(i.VoiceStates)
	}
}

```

Thank you for helping.

**Reply by topi314** [reply]
ready is sent before all guilds are sent
check them  in `GuildCreate`

**Reply by StateGovernment** [reply]
@TopiSenpai @FedorLap2006 #1021 This issue has the same details attached, and it worked out for me. Thank you for the help.

For future readers, You've gotta add both ready event and guildCreate event inorder to get voicestates at a moment.

My only last question would be, How often does Guildcreate be called, does it only initiate when we open a connection or does it have a timeInterval at which it triggers?

**Reply by topi314** [reply]
first mark your question as answered
second for more infos on when guild creates are fired checkout the discord docs https://discord.com/developers/docs/topics/gateway#guild-create
